frame/preimage/src/lib.rs

```diff
@@ -0,0 +1,347 @@
+	/// There are a non-zero number of outstanding requests for this hash by this chain. If there
+	/// is a preimage registered, then it may be removed iff this counter becomes zero.
+	Requested(u32),
+}
+
+type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The Weight information for this pallet.
+		type WeightInfo: weights::WeightInfo;
+
+		/// Currency type for this pallet.
+		type Currency: ReservableCurrency<Self::AccountId>;
+
+		/// An origin that can request a preimage be placed on-chain without a deposit or fee, or
+		/// manage existing preimages.
+		type ManagerOrigin: EnsureOrigin<Self::Origin>;
+
+		/// Max size allowed for a preimage.
+		type MaxSize: Get<u32>;
+
+		/// The base deposit for placing a preimage on chain.
+		type BaseDeposit: Get<BalanceOf<Self>>;
+
+		/// The per-byte deposit for placing a preimage on chain.
+		type ByteDeposit: Get<BalanceOf<Self>>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::generate_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A preimage has been noted.
+		Noted { hash: T::Hash },
+		/// A preimage has been requested.
+		Requested { hash: T::Hash },
+		/// A preimage has ben cleared.
+		Cleared { hash: T::Hash },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Preimage is too large to store on-chain.
+		TooLarge,
+		/// Preimage has already been noted on-chain.
+		AlreadyNoted,
+		/// The user is not authorized to perform this action.
+		NotAuthorized,
+		/// The preimage cannot be removed since it has not yet been noted.
+		NotNoted,
+		/// A preimage may not be removed when there are outstanding requests.
+		Requested,
+		/// The preimage request cannot be removed since no outstanding requests exist.
+		NotRequested,
+	}
+
+	/// The request status of a given hash.
+	#[pallet::storage]
+	pub(super) type StatusFor<T: Config> =
+		StorageMap<_, Identity, T::Hash, RequestStatus<T::AccountId, BalanceOf<T>>>;
+
+	/// The preimages stored by this pallet.
+	#[pallet::storage]
+	pub(super) type PreimageFor<T: Config> =
+		StorageMap<_, Identity, T::Hash, BoundedVec<u8, T::MaxSize>>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Register a preimage on-chain.
+		///
+		/// If the preimage was previously requested, no fees or deposits are taken for providing
+		/// the preimage. Otherwise, a deposit is taken proportional to the size of the preimage.
+		#[pallet::weight(T::WeightInfo::note_preimage(bytes.len() as u32))]
+		pub fn note_preimage(origin: OriginFor<T>, bytes: Vec<u8>) -> DispatchResultWithPostInfo {
```

Comment:
why don't we already make this be a `BoundedVec` and let the decoding length fail in earlier stages? 
 
```suggestion
		pub fn note_preimage(origin: OriginFor<T>, bytes: BoundedVec<u8>) -> DispatchResultWithPostInfo {
```

Reply:
I think APIs should use generic types rather than custom ones. This would, for example, cause us to break the tx version everytime we update what the bound of the vec would be on this pallet.